From: =?utf-8?q?Timo_R=C3=B6hling?= <timo.roehling@fkie.fraunhofer.de>
Date: Sat, 11 Apr 2020 20:45:02 +0200
Subject: Add pkg-config support

This PR implements the wishlist bug from #22. The pkg-config file format
is pretty simple, it basically provides a bunch of compiler flags to
compile and link against a library.

I changed the default INSTALL_DIR locations to relative paths, because
CMake expands relative paths to the CMAKE_INSTALL_PREFIX by default
anyway, and it makes it possible to have both the CMake config and the
pkg-config in a relocatable style.

Most of the magic happens in the foreach() loop, which is a semicolon
separated list of config values for the qhull.pc.in template in the form
NAME;DESCRIPTION
---
 CMakeLists.txt | 29 ++++++++++++++++++++++-------
 qhull.pc.in    | 10 ++++++++++
 2 files changed, 32 insertions(+), 7 deletions(-)
 create mode 100644 qhull.pc.in

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 080d764..cfc2269 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -81,30 +81,30 @@ check_lfs(WITH_LFS)
 
 if(INCLUDE_INSTALL_DIR)
 else()
-set(INCLUDE_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/include)
+set(INCLUDE_INSTALL_DIR include)
 endif()
 if(LIB_INSTALL_DIR)
 else()
-set(LIB_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/lib)
+set(LIB_INSTALL_DIR lib)
 endif()
 if(BIN_INSTALL_DIR)
 else()
-set(BIN_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/bin)
+set(BIN_INSTALL_DIR bin)
 endif()
 if(MAN_INSTALL_DIR)
 else()
     if(WIN32)
-        set(MAN_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/man/man1)
+        set(MAN_INSTALL_DIR man/man1)
     else()
-        set(MAN_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/share/man/man1)
+        set(MAN_INSTALL_DIR share/man/man1)
     endif()
 endif()
 if(DOC_INSTALL_DIR)
 else()
     if(WIN32)
-        set(DOC_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/doc)
+        set(DOC_INSTALL_DIR doc)
     else()
-        set(DOC_INSTALL_DIR ${CMAKE_INSTALL_PREFIX}/share/doc/qhull)
+        set(DOC_INSTALL_DIR share/doc/qhull)
     endif()
 endif()
 message(STATUS)
@@ -662,6 +662,21 @@ install(
     Devel
 )
 
+set(PkgConfigLocation lib/pkgconfig)
+foreach(pkgconfig IN ITEMS "${qhull_SHARED};Qhull shared library"
+                           "${qhull_SHAREDR};Qhull reentrant shared library"
+                           "${qhull_STATIC};Qhull static library"
+                           "${qhull_STATICR};Qhull reentrant static library"
+                           "${qhull_CPP};Qhull C++ library")
+    list(GET pkgconfig 0 LIBRARY_NAME)
+    if(LIBRARY_NAME STREQUAL "libqhull")
+        set(LIBRARY_NAME "qhull")
+    endif()
+    list(GET pkgconfig 1 LIBRARY_DESCRIPTION)
+    configure_file(qhull.pc.in ${LIBRARY_NAME}.pc @ONLY)
+    install(FILES ${CMAKE_CURRENT_BINARY_DIR}/${LIBRARY_NAME}.pc DESTINATION ${PkgConfigLocation} COMPONENT Devel)
+endforeach()
+
 install(FILES ${libqhull_HEADERS}    DESTINATION ${INCLUDE_INSTALL_DIR}/libqhull)
 install(FILES ${libqhull_DOC}        DESTINATION ${INCLUDE_INSTALL_DIR}/libqhull)
 install(FILES ${libqhullr_HEADERS}    DESTINATION ${INCLUDE_INSTALL_DIR}/libqhull_r)
diff --git a/qhull.pc.in b/qhull.pc.in
new file mode 100644
index 0000000..f4d7b78
--- /dev/null
+++ b/qhull.pc.in
@@ -0,0 +1,10 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+libdir=${prefix}/@LIB_INSTALL_DIR@
+includedir=${prefix}/@INCLUDE_INSTALL_DIR@
+
+Name: @LIBRARY_NAME@
+Description: @LIBRARY_DESCRIPTION@
+Version: @qhull_VERSION@
+Libs: -L${libdir} -l@LIBRARY_NAME@
+Cflags: -I${includedir}
+
